src/lib.rs

```diff
@@ -304,21 +311,26 @@ where
         }
         let time = iter_start.elapsed();
         data.push((iters, time));
-    }
 
-    // If the first iter in a sample is consistently slow, that's fine -
-    // that's why we do the linear regression. If the first sample is slower
-    // than the rest, however, that's not fine.  Therefore, we discard the
-    // first sample as a cache-warming exercise.
-    data.remove(0);
-
-    // Compute some stats
-    let (grad, r2) = regression(&data[..]);
-    Stats {
-        ns_per_iter: grad,
-        goodness_of_fit: r2,
-        iterations: data.iter().map(|&(x, _)| x).sum(),
-        samples: data.len(),
+        let elapsed = bench_start.elapsed();
+        if elapsed > BENCH_TIME_MIN && data.len() > 3 {
+            // If the first iter in a sample is consistently slow, that's fine -
+            // that's why we do the linear regression. If the first sample is slower
+            // than the rest, however, that's not fine.  Therefore, we discard the
+            // first sample as a cache-warming exercise.
+
+            // Compute some stats
+            let (grad, r2) = regression(&data[1..]);
+            let stats = Stats {
+                ns_per_iter: grad,
+                goodness_of_fit: r2,
+                iterations: data[1..].iter().map(|&(x, _)| x).sum(),
+                samples: data[1..].len(),
+            };
+            if elapsed > BENCH_TIME_MAX || r2 > 0.99 {
+                return stats;
+            }
```

Comment:
Note that R2 is not like a confidence interval, which gets tighter as you add more measurements.  If the machine is noisy then the R2 will be bad, no matter how many measurements you take.  On the other hand, if the machine gives you a bad measurement which throws off the R2, and then goes back to producing consistent results, you can eventually drown out the bad one and make the R2 good again.  I guess I'm saying that the utility of approach depends on the kind of noise the benchmarking machine experiences.

Reply:
Yeah, I've observed that.  So the upper time limit is still very important, as a noisy computation will always run to the upper limit, and it might be nice to have a lower upper limit than 10s.

Do you think it would be worth actually trying to estimate the error on the slope? The trouble (as I see it) is that it is hard to distinguish between noise and a systematic issue.  My best guess for an error estimate would be to do a fit to a parabola and a fit to the straight line and compare the linear terms.  If they differ then presumably we want more data?

Reply:
One note on this that I realized.  In a general case R2 doesn't get tighter as you add more measurements, but in our case I believe R2 should drop as we add more data, provided there are only two sources of error:  (1) an error on each timing, which is independent of the time, and (2) a fluctuation in the time required by a given iteration, which has a mean value.  If there isn't some error that increases with N (the number of iterations), then our fractional error in the total time should drop as we increase the number of iterations, and R2 should drop as well.

Not that R2 is easily (or accurately) interpreted as an uncertainty, but I do think that it should always approach 1 as we collect more statistics, as long as we keep adding larger N runs, and never repeat the smaller N runs.